pkg/util/validation/override.go

```diff
@@ -22,290 +19,87 @@ func init() {
+func NewOverridesManager(cfg OverridesManagerConfig) (*OverridesManager, error) {
+	overridesManager := OverridesManager{
+		cfg: cfg,
 	}
 
-	o := &Overrides{
-		Defaults:  defaults,
-		overrides: overrides,
-		quit:      make(chan struct{}),
+	if cfg.OverridesLoadPath != "" {
+		overridesManager.loop()
+	} else {
+		level.Info(util.Logger).Log("msg", "per-tenant overrides disabled")
 	}
 
-	go o.loop()
-	return o, nil
+	return &overridesManager, nil
 }
 
-func (o *Overrides) loop() {
-	ticker := time.NewTicker(o.Defaults.PerTenantOverridePeriod)
+func (om *OverridesManager) loop() {
+	ticker := time.NewTicker(om.cfg.OverridesReloadPeriod)
 	defer ticker.Stop()
 
 	for {
 		select {
 		case <-ticker.C:
-			overrides, err := loadOverrides(o.Defaults.PerTenantOverrideConfig)
+			err := om.loadOverrides()
 			if err != nil {
-				overridesReloadSuccess.Set(0)
-				level.Error(util.Logger).Log("msg", "failed to reload overrides", "err", err)
-				continue
+				level.Error(util.Logger).Log("msg", "failed to load limit overrides", "err", err)
 			}
-			overridesReloadSuccess.Set(1)
-
-			o.overridesMtx.Lock()
-			o.overrides = overrides
-			o.overridesMtx.Unlock()
-		case <-o.quit:
+		case <-om.quit:
 			return
 		}
 	}
 }
 
-// Stop background reloading of overrides.
-func (o *Overrides) Stop() {
-	close(o.quit)
-}
-
-func loadOverrides(filename string) (map[string]*Limits, error) {
-	f, err := os.Open(filename)
+func (om *OverridesManager) loadOverrides() error {
+	overrides, err := om.cfg.OverridesLoader(om.cfg.OverridesLoadPath)
 	if err != nil {
-		return nil, err
-	}
-
-	var overrides struct {
-		Overrides map[string]*Limits `yaml:"overrides"`
-	}
-
-	decoder := yaml.NewDecoder(f)
-	decoder.SetStrict(true)
-	if err := decoder.Decode(&overrides); err != nil {
-		return nil, err
+		overridesReloadSuccess.Set(0)
+		return err
 	}
+	overridesReloadSuccess.Set(1)
 
-	return overrides.Overrides, nil
+	om.overridesMtx.Lock()
+	defer om.overridesMtx.Unlock()
+	om.overrides = overrides
+	return nil
 }
 
-func (o *Overrides) getBool(userID string, f func(*Limits) bool) bool {
-	o.overridesMtx.RLock()
-	defer o.overridesMtx.RUnlock()
-	override, ok := o.overrides[userID]
-	if !ok {
-		return f(&o.Defaults)
-	}
-	return f(override)
+// Stop stops the OverridesManager
+func (om *OverridesManager) Stop() {
+	close(om.quit)
 }
 
-func (o *Overrides) getInt(userID string, f func(*Limits) int) int {
-	o.overridesMtx.RLock()
-	defer o.overridesMtx.RUnlock()
-	override, ok := o.overrides[userID]
-	if !ok {
-		return f(&o.Defaults)
-	}
-	return f(override)
-}
-
-func (o *Overrides) getFloat(userID string, f func(*Limits) float64) float64 {
-	o.overridesMtx.RLock()
-	defer o.overridesMtx.RUnlock()
-	override, ok := o.overrides[userID]
-	if !ok {
-		return f(&o.Defaults)
-	}
-	return f(override)
-}
+// GetLimits returns Limits for a specific userID if its set otherwise the default Limits
+func (om *OverridesManager) GetLimits(userID string) interface{} {
```

Comment:
What happends when the overrides are only partially specified?

ie if defaults are:
```yaml
a: 1
b: 2
```

and the overrides are:
```yaml
foo:
  b: 3
```

`GetLimits("foo")` will return:
```yaml
a: 0
b: 3
```
not
```yaml
a: 1
b: 3
```

Reply:
Not it won't ignore me.